lib/class-wp-annotation-utils.php

```diff
@@ -0,0 +1,1074 @@
+			return $error;
+		}
+
+		foreach ( $selection['ranges'] as $range ) {
+			if ( ! is_array( $range ) ) {
+				return $error;
+			} elseif ( ! isset( $range['begin']['offset'], $range['end']['offset'] ) ) {
+				return $error;
+			} elseif ( ! is_int( $range['begin']['offset'] ) || ! is_int( $range['end']['offset'] ) ) {
+				return $error;
+			} elseif ( 2 !== count( array_keys( $range ) ) ) {
+				return $error;
+			}
+		}
+
+		return $selection;
+	}
+
+	/**
+	 * Validates a substatus update.
+	 *
+	 * @since [version]
+	 * @access protected
+	 *
+	 * @param  string          $substatus Substatus.
+	 * @param  WP_Post         $post      Post (annotation).
+	 * @param  WP_Rest_Request $request   REST API request.
+	 * @return string|WP_Error            Substatus, else WP_Error on validation failure.
+	 *
+	 * @see on_update_additional_rest_field()
+	 */
+	protected static function validate_rest_field_substatus_on_update( $substatus, $post, $request ) {
+		if ( ! in_array( $substatus, self::$substatuses, true ) ) {
+			return self::rest_field_validation_update_error( 'substatus' );
+		}
+
+		return $substatus;
+	}
+
+	/**
+	 * Maybe update substatus history following a substatus change.
+	 *
+	 * @since [version]
+	 * @access protected
+	 *
+	 * @param  string          $new     New substatus.
+	 * @param  string          $old     Old substatus.
+	 * @param  WP_Post         $post    Post (annotation).
+	 * @param  WP_Rest_Request $request REST API request.
+	 *
+	 * @see on_update_additional_rest_field()
+	 */
+	protected static function maybe_update_rest_field_substatus_history( $new, $old, $post, $request ) {
+		if ( $new === $old ) {
+			return; // No change.
+		}
+
+		$current_time      = time();
+		$user              = wp_get_current_user();
+		$new_history_entry = array(); // Initialize.
+
+		$history = get_post_meta( $post->ID, '_substatus_history', true );
+		$history = is_array( $history ) ? $history : array();
+
+		$annotator = self::validate_rest_field_annotator_on_update(
+			$request['annotator'], $request['annotator_meta'], $post, $request
+		);
+
+		if ( ! is_wp_error( $annotator ) ) {
+			$new_history_entry = array(
+				'identity'      => $annotator['id'],
+				'identity_meta' => array(
+					'display_name' => $annotator['meta']['display_name'],
+					'md5_email'    => $annotator['meta']['md5_email'],
+				),
+				'time'          => $current_time,
+				'old'           => $old,
+				'new'           => $new,
+			);
+		} elseif ( $user->exists() ) {
+			$new_history_entry = array(
+				'identity'      => (string) $user->ID,
+				'identity_meta' => array(
+					'display_name' => $user->display_name,
+					'md5_email'    => md5( strtolower( $user->user_email ) ),
+				),
+				'time'          => $current_time,
+				'old'           => $old,
+				'new'           => $new,
+			);
+		}
+
+		if ( $new_history_entry ) {
+			$history[] = $new_history_entry;
+			$history   = array_slice( $history, -25 ); // Last 25 changes only.
```

Comment:
Same as in my other comment. Is it possible to load "_25_" from some global configuration instead? It would help with maintenance and synchronisation of this class with whatever else is going on.

Reply:
I think it might be good to do an `apply_filters` on this.